arraycontext/container/traversal.py

```diff
@@ -279,6 +291,52 @@ def rec(keys: Tuple[Union[str, int], ...],
 # }}}
 
 
+# {{{ array container reductions
+
+def rec_reduce_array_container(
+        reduce_func: Callable[[Iterable[Any]], Any],
+        array_func: Callable[[Any], Any],
+        ary: ArrayContainerT) -> Any:
+    """Perform reductions over array containers recursively.
+
+    :param reduce_func: callable used to reduce over the components of the
+        :class:`~arraycontext.ArrayContainer`.
+    :param array_func: callable used to reduce a single component of the
+        :class:`~arraycontext.ArrayContainer`. The callable takes arrays of
+        type :class:`arraycontext.ArrayContext.array_types` and returns an
+        array of the same type or a scalar.
+    """
+    def _reduce_wrapper(ary: Any, iterable: Iterable[Tuple[Any, Any]]) -> Any:
+        return reduce_func([subary for _, subary in iterable])
```

Comment:
Keep this a generator to avoid memory allocation?

Reply:
I'm on the side of keeping it a list too (only used in one place now too). How strongly do you feel about this?

Reply:
Was a minor suggestion. But looks like generators have their downsides too, that was overlooked. Let's resolve it.